compiler-rt/lib/builtins/aarch64/sme-abi.S

```diff
@@ -204,6 +206,161 @@ DEFINE_COMPILERRT_FUNCTION(__arm_get_current_vg)
   ret
 END_COMPILERRT_FUNCTION(__arm_get_current_vg)
 
+DEFINE_COMPILERRT_FUNCTION(__arm_sme_state_size)
+  .variant_pcs __arm_sme_state_size
+  BTI_C
+
+  // Test if SME is available and PSTATE = 1.
+  adrp    x16, CPU_FEATS_SYMBOL
+  ldr     x16, [x16, CPU_FEATS_SYMBOL_OFFSET]
+  tbz     x16, #FEAT_SME_BIT, 0f
+  mrs     x16, SVCR
+  tbz     x16, #1, 0f
```

Comment:
Does this routine also need to check that TPIDR2_EL0 is null?

Reply:
From the specification:
> If the ZA state is active, X0 contains the total size required to save and restore all SME state enabled by PSTATE.ZA.

So yes, there's no need to allocate more size than necessary. Good catch!